File: server/src/group/group.controller.ts
import {
  Controller,
  Get,
  Post,
  Put,
  Delete,
  Body,
  Param,
  Query,
  UseGuards,
  HttpException,
  HttpStatus,
} from '@nestjs/common';
import {
  ApiTags,
  ApiOperation,
  ApiResponse,
  ApiBearerAuth,
} from '@nestjs/swagger';
import { GroupService } from './group.service';
import { AuthGuard } from '../auth/guards/auth.guard';
import { CurrentUserId } from '../auth/decorators/current-user.decorator';
import {
  CreateGroupDto,
  UpdateGroupDto,
  GetGroupsQueryDto,
  UpdateMemberRoleDto,
} from './dto/group.dto';

/**
 * 그룹 관리 컨트롤러
 */
@ApiTags('groups')
@Controller('groups')
@UseGuards(AuthGuard)
@ApiBearerAuth('JWT-auth')
export class GroupController {
  constructor(private readonly groupService: GroupService) {}

  /**
   * 그룹 목록 조회
   */
  @Get()
  @ApiOperation({ summary: '그룹 목록 조회' })
  @ApiResponse({ status: 200, description: '그룹 목록 조회 성공' })
  async getGroups(
    @CurrentUserId() userId: string,
    @Query() query: GetGroupsQueryDto,
  ) {
    try {
      const groups = await this.groupService.getGroups(
        userId,
        query.type,
        query.search,
        query.page || 1,
        query.limit || 20,
      );
      return {
        success: true,
        data: groups,
      };
    } catch (error) {
      throw new HttpException(
        {
          success: false,
          message: error.message || '그룹 목록 조회에 실패했습니다.',
        },
        error.status || HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }

  /**
   * 새 그룹 생성
   */
  @Post()
  @ApiOperation({ summary: '새 그룹 생성' })
  @ApiResponse({ status: 201, description: '그룹 생성 성공' })
  @ApiResponse({ status: 400, description: '잘못된 요청' })
  async createGroup(
    @CurrentUserId() userId: string,
    @Body() createGroupDto: CreateGroupDto,
  ) {
    try {
      const group = await this.groupService.createGroup(userId, createGroupDto);
      return {
        success: true,
        data: group,
      };
    } catch (error) {
      throw new HttpException(
        {
          success: false,
          message: error.message || '그룹 생성에 실패했습니다.',
        },
        error.status || HttpStatus.BAD_REQUEST,
      );
    }
  }

  /**
   * 특정 그룹 조회
   */
  @Get(':id')
  @ApiOperation({ summary: '특정 그룹 조회' })
  @ApiResponse({ status: 200, description: '그룹 조회 성공' })
  @ApiResponse({ status: 404, description: '그룹을 찾을 수 없음' })
  async getGroupById(
    @CurrentUserId() userId: string,
    @Param('id') groupId: string,
  ) {
    try {
      const group = await this.groupService.getGroupById(groupId, userId);
      return {
        success: true,
        data: group,
      };
    } catch (error) {
      throw new HttpException(
        {
          success: false,
          message: error.message || '그룹 조회에 실패했습니다.',
        },
        error.status || HttpStatus.NOT_FOUND,
      );
    }
  }

  /**
   * 그룹 정보 수정
   */
  @Put(':id')
  @ApiOperation({ summary: '그룹 정보 수정' })
  @ApiResponse({ status: 200, description: '그룹 수정 성공' })
  @ApiResponse({ status: 403, description: '권한 없음' })
  async updateGroup(
    @CurrentUserId() userId: string,
    @Param('id') groupId: string,
    @Body() updateGroupDto: UpdateGroupDto,
  ) {
    try {
      const group = await this.groupService.updateGroup(
        groupId,
        userId,
        updateGroupDto,
      );
      return {
        success: true,
        data: group,
      };
    } catch (error) {
      throw new HttpException(
        {
          success: false,
          message: error.message || '그룹 수정에 실패했습니다.',
        },
        error.status || HttpStatus.FORBIDDEN,
      );
    }
  }

  // 그룹 삭제 기능은 서비스에 구현되지 않음 - 필요 시 추가 구현 필요

  /**
   * 그룹 참여
   */
  @Post(':id/join')
  @ApiOperation({ summary: '그룹 참여' })
  @ApiResponse({ status: 200, description: '그룹 참여 성공' })
  @ApiResponse({ status: 400, description: '참여 실패' })
  async joinGroup(
    @CurrentUserId() userId: string,
    @Param('id') groupId: string,
  ) {
    try {
      await this.groupService.joinGroup(groupId, userId);
      return {
        success: true,
        message: '그룹에 참여했습니다.',
      };
    } catch (error) {
      throw new HttpException(
        {
          success: false,
          message: error.message || '그룹 참여에 실패했습니다.',
        },
        error.status || HttpStatus.BAD_REQUEST,
      );
    }
  }

  /**
   * 그룹 탈퇴
   */
  @Delete(':id/leave')
  @ApiOperation({ summary: '그룹 탈퇴' })
  @ApiResponse({ status: 200, description: '그룹 탈퇴 성공' })
  async leaveGroup(
    @CurrentUserId() userId: string,
    @Param('id') groupId: string,
  ) {
    try {
      await this.groupService.leaveGroup(groupId, userId);
      return {
        success: true,
        message: '그룹에서 탈퇴했습니다.',
      };
    } catch (error) {
      throw new HttpException(
        {
          success: false,
          message: error.message || '그룹 탈퇴에 실패했습니다.',
        },
        error.status || HttpStatus.BAD_REQUEST,
      );
    }
  }

  /**
   * 사용자의 그룹 목록 조회
   */
  @Get('my-groups')
  @ApiOperation({ summary: '내 그룹 목록 조회' })
  @ApiResponse({ status: 200, description: '그룹 목록 조회 성공' })
  async getMyGroups(@CurrentUserId() userId: string) {
    try {
      const groups = await this.groupService.getUserGroups(userId);
      return {
        success: true,
        data: groups,
      };
    } catch (error) {
      throw new HttpException(
        {
          success: false,
          message: error.message || '그룹 목록 조회에 실패했습니다.',
        },
        error.status || HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }

  /**
   * 초대 링크 생성
   */
  @Post(':id/invites')
  @ApiOperation({ summary: '초대 링크 생성' })
  @ApiResponse({ status: 201, description: '초대 링크 생성 성공' })
  async generateInviteLink(
    @CurrentUserId() userId: string,
    @Param('id') groupId: string,
  ) {
    try {
      const inviteLink = await this.groupService.generateInviteLink(
        groupId,
        userId,
      );
      return {
        success: true,
        data: { inviteLink },
      };
    } catch (error) {
      throw new HttpException(
        {
          success: false,
          message: error.message || '초대 링크 생성에 실패했습니다.',
        },
        error.status || HttpStatus.BAD_REQUEST,
      );
    }
  }

  /**
   * 초대 코드로 그룹 참여
   */
  @Post('join/:code')
  @ApiOperation({ summary: '초대 코드로 그룹 참여' })
  @ApiResponse({ status: 200, description: '그룹 참여 성공' })
  async joinGroupByInvite(
    @CurrentUserId() userId: string,
    @Param('code') inviteCode: string,
  ) {
    try {
      const result = await this.groupService.joinByInviteCode(
        inviteCode,
        userId,
      );
      return {
        success: true,
        data: result,
      };
    } catch (error) {
      throw new HttpException(
        {
          success: false,
          message: error.message || '초대 코드를 사용한 참여에 실패했습니다.',
        },
        error.status || HttpStatus.BAD_REQUEST,
      );
    }
  }

  /**
   * 멤버 역할 변경
   */
  @Put(':id/members/:memberId')
  @ApiOperation({ summary: '멤버 역할 변경' })
  @ApiResponse({ status: 200, description: '역할 변경 성공' })
  async updateMemberRole(
    @CurrentUserId() userId: string,
    @Param('id') groupId: string,
    @Param('memberId') memberId: string,
    @Body() updateMemberRoleDto: UpdateMemberRoleDto,
  ) {
    try {
      await this.groupService.updateMemberRole(
        groupId,
        userId,
        memberId,
        updateMemberRoleDto,
      );
      return {
        success: true,
        message: '멤버 역할이 변경되었습니다.',
      };
    } catch (error) {
      throw new HttpException(
        {
          success: false,
          message: error.message || '멤버 역할 변경에 실패했습니다.',
        },
        error.status || HttpStatus.FORBIDDEN,
      );
    }
  }

  /**
   * 멤버 제거
   */
  @Delete(':id/members/:memberId')
  @ApiOperation({ summary: '멤버 제거' })
  @ApiResponse({ status: 200, description: '멤버 제거 성공' })
  async removeMember(
    @CurrentUserId() userId: string,
    @Param('id') groupId: string,
    @Param('memberId') memberId: string,
  ) {
    try {
      await this.groupService.removeMember(groupId, userId, memberId);
      return {
        success: true,
        message: '멤버가 제거되었습니다.',
      };
    } catch (error) {
      throw new HttpException(
        {
          success: false,
          message: error.message || '멤버 제거에 실패했습니다.',
        },
        error.status || HttpStatus.FORBIDDEN,
      );
    }
  }
}

File: server2/src/main/java/com/glimpse/server/controller/GroupController.java
package com.glimpse.server.controller;

import com.glimpse.server.dto.common.ApiResponse;
import com.glimpse.server.dto.group.CreateGroupDto;
import com.glimpse.server.dto.group.GroupDto;
import com.glimpse.server.dto.group.UpdateGroupDto;
import com.glimpse.server.entity.enums.GroupCategory;
import com.glimpse.server.entity.enums.GroupType;
import com.glimpse.server.service.GroupService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

/**
 * Group Controller
 *
 * <p>그룹 관련 REST API를 제공하는 컨트롤러입니다.</p>
 *
 * @author Glimpse Team
 * @version 1.0
 * @since 2025-01-14
 */
@Slf4j
@RestController
@RequestMapping("/api/v1/groups")
@RequiredArgsConstructor
@Tag(name = "Group", description = "그룹 관리 API")
public class GroupController {

    private final GroupService groupService;

    @PostMapping
    @Operation(summary = "그룹 생성", description = "새로운 그룹을 생성합니다")
    public ResponseEntity<ApiResponse<GroupDto>> createGroup(
            @Valid @RequestBody CreateGroupDto createGroupDto,
            @RequestParam String creatorId) {
        log.info("POST /api/v1/groups - Creating group: {} by user: {}", createGroupDto.getName(), creatorId);
        try {
            GroupDto groupDto = groupService.createGroup(createGroupDto, creatorId);
            ApiResponse<GroupDto> response = ApiResponse.<GroupDto>builder()
                    .success(true)
                    .data(groupDto)
                    .message("그룹이 성공적으로 생성되었습니다")
                    .build();
            return ResponseEntity.status(HttpStatus.CREATED).body(response);
        } catch (IllegalArgumentException e) {
            log.error("Failed to create group: {}", e.getMessage());
            ApiResponse<GroupDto> response = ApiResponse.<GroupDto>builder()
                    .success(false)
                    .message(e.getMessage())
                    .build();
            return ResponseEntity.badRequest().body(response);
        }
    }

    @GetMapping("/{id}")
    @Operation(summary = "그룹 조회", description = "ID로 그룹을 조회합니다")
    public ResponseEntity<ApiResponse<GroupDto>> getGroupById(@PathVariable String id) {
        log.info("GET /api/v1/groups/{} - Getting group", id);
        return groupService.getGroupById(id)
                .map(group -> {
                    ApiResponse<GroupDto> response = ApiResponse.<GroupDto>builder()
                            .success(true)
                            .data(group)
                            .build();
                    return ResponseEntity.ok(response);
                })
                .orElseGet(() -> {
                    ApiResponse<GroupDto> response = ApiResponse.<GroupDto>builder()
                            .success(false)
                            .message("그룹을 찾을 수 없습니다: " + id)
                            .build();
                    return ResponseEntity.status(HttpStatus.NOT_FOUND).body(response);
                });
    }

    @PatchMapping("/{id}")
    @Operation(summary = "그룹 업데이트", description = "그룹 정보를 업데이트합니다")
    public ResponseEntity<ApiResponse<GroupDto>> updateGroup(
            @PathVariable String id,
            @Valid @RequestBody UpdateGroupDto updateGroupDto) {
        log.info("PATCH /api/v1/groups/{} - Updating group", id);
        try {
            GroupDto groupDto = groupService.updateGroup(id, updateGroupDto);
            ApiResponse<GroupDto> response = ApiResponse.<GroupDto>builder()
                    .success(true)
                    .data(groupDto)
                    .message("그룹 정보가 업데이트되었습니다")
                    .build();
            return ResponseEntity.ok(response);
        } catch (IllegalArgumentException e) {
            log.error("Failed to update group: {}", e.getMessage());
            ApiResponse<GroupDto> response = ApiResponse.<GroupDto>builder()
                    .success(false)
                    .message(e.getMessage())
                    .build();
            return ResponseEntity.badRequest().body(response);
        }
    }

    @DeleteMapping("/{id}")
    @Operation(summary = "그룹 삭제", description = "그룹을 삭제합니다")
    public ResponseEntity<ApiResponse<Void>> deleteGroup(@PathVariable String id) {
        log.info("DELETE /api/v1/groups/{} - Deleting group", id);
        try {
            groupService.deleteGroup(id);
            ApiResponse<Void> response = ApiResponse.<Void>builder()
                    .success(true)
                    .message("그룹이 삭제되었습니다")
                    .build();
            return ResponseEntity.ok(response);
        } catch (IllegalArgumentException e) {
            log.error("Failed to delete group: {}", e.getMessage());
            ApiResponse<Void> response = ApiResponse.<Void>builder()
                    .success(false)
                    .message(e.getMessage())
                    .build();
            return ResponseEntity.badRequest().body(response);
        }
    }

    @GetMapping
    @Operation(summary = "공개 그룹 목록 조회", description = "모든 공개 활성 그룹을 조회합니다")
    public ResponseEntity<ApiResponse<List<GroupDto>>> getPublicGroups(
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "20") int size) {
        log.info("GET /api/v1/groups - Getting public groups (page: {}, size: {})", page, size);
        
        if (page == 0 && size == 20) {
            // 페이징 없이 전체 조회
            List<GroupDto> groups = groupService.getPublicGroups();
            ApiResponse<List<GroupDto>> response = ApiResponse.<List<GroupDto>>builder()
                    .success(true)
                    .data(groups)
                    .build();
            return ResponseEntity.ok(response);
        } else {
            // 페이징 처리
            Pageable pageable = PageRequest.of(page, size);
            Page<GroupDto> groupPage = groupService.getPublicGroupsPage(pageable);
            ApiResponse<List<GroupDto>> response = ApiResponse.<List<GroupDto>>builder()
                    .success(true)
                    .data(groupPage.getContent())
                    .build();
            return ResponseEntity.ok(response);
        }
    }

    @GetMapping("/official")
    @Operation(summary = "공식 그룹 목록 조회", description = "모든 공식 그룹을 조회합니다")
    public ResponseEntity<ApiResponse<List<GroupDto>>> getOfficialGroups() {
        log.info("GET /api/v1/groups/official - Getting official groups");
        List<GroupDto> groups = groupService.getOfficialGroups();
        ApiResponse<List<GroupDto>> response = ApiResponse.<List<GroupDto>>builder()
                .success(true)
                .data(groups)
                .build();
        return ResponseEntity.ok(response);
    }

    @GetMapping("/type/{type}")
    @Operation(summary = "타입별 그룹 조회", description = "특정 타입의 그룹을 조회합니다")
    public ResponseEntity<ApiResponse<List<GroupDto>>> getGroupsByType(@PathVariable GroupType type) {
        log.info("GET /api/v1/groups/type/{} - Getting groups by type", type);
        List<GroupDto> groups = groupService.getGroupsByType(type);
        ApiResponse<List<GroupDto>> response = ApiResponse.<List<GroupDto>>builder()
                .success(true)
                .data(groups)
                .build();
        return ResponseEntity.ok(response);
    }

    @GetMapping("/category/{category}")
    @Operation(summary = "카테고리별 그룹 조회", description = "특정 카테고리의 그룹을 조회합니다")
    public ResponseEntity<ApiResponse<List<GroupDto>>> getGroupsByCategory(@PathVariable GroupCategory category) {
        log.info("GET /api/v1/groups/category/{} - Getting groups by category", category);
        List<GroupDto> groups = groupService.getGroupsByCategory(category);
        ApiResponse<List<GroupDto>> response = ApiResponse.<List<GroupDto>>builder()
                .success(true)
                .data(groups)
                .build();
        return ResponseEntity.ok(response);
    }

    @GetMapping("/creator/{userId}")
    @Operation(summary = "사용자가 생성한 그룹 조회", description = "특정 사용자가 생성한 그룹을 조회합니다")
    public ResponseEntity<ApiResponse<List<GroupDto>>> getGroupsByCreator(@PathVariable String userId) {
        log.info("GET /api/v1/groups/creator/{} - Getting groups by creator", userId);
        List<GroupDto> groups = groupService.getGroupsByCreator(userId);
        ApiResponse<List<GroupDto>> response = ApiResponse.<List<GroupDto>>builder()
                .success(true)
                .data(groups)
                .build();
        return ResponseEntity.ok(response);
    }

    @GetMapping("/user/{userId}")
    @Operation(summary = "사용자가 가입한 그룹 조회", description = "특정 사용자가 가입한 그룹을 조회합니다")
    public ResponseEntity<ApiResponse<List<GroupDto>>> getGroupsByUserId(@PathVariable String userId) {
        log.info("GET /api/v1/groups/user/{} - Getting groups by user", userId);
        List<GroupDto> groups = groupService.getGroupsByUserId(userId);
        ApiResponse<List<GroupDto>> response = ApiResponse.<List<GroupDto>>builder()
                .success(true)
                .data(groups)
                .build();
        return ResponseEntity.ok(response);
    }

    @GetMapping("/search")
    @Operation(summary = "그룹 검색", description = "키워드로 그룹을 검색합니다")
    public ResponseEntity<ApiResponse<List<GroupDto>>> searchGroups(
            @RequestParam String keyword,
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "20") int size) {
        log.info("GET /api/v1/groups/search?keyword={} - Searching groups", keyword);
        Pageable pageable = PageRequest.of(page, size);
        Page<GroupDto> groupPage = groupService.searchGroups(keyword, pageable);
        ApiResponse<List<GroupDto>> response = ApiResponse.<List<GroupDto>>builder()
                .success(true)
                .data(groupPage.getContent())
                .build();
        return ResponseEntity.ok(response);
    }

    @GetMapping("/nearby")
    @Operation(summary = "주변 그룹 조회", description = "위치 기반으로 주변 그룹을 조회합니다")
    public ResponseEntity<ApiResponse<List<GroupDto>>> getNearbyGroups(
            @RequestParam double latitude,
            @RequestParam double longitude,
            @RequestParam(defaultValue = "10.0") double radiusKm) {
        log.info("GET /api/v1/groups/nearby?lat={}&lng={}&radius={}", latitude, longitude, radiusKm);
        List<GroupDto> groups = groupService.getNearbyGroups(latitude, longitude, radiusKm);
        ApiResponse<List<GroupDto>> response = ApiResponse.<List<GroupDto>>builder()
                .success(true)
                .data(groups)
                .build();
        return ResponseEntity.ok(response);
    }

    @GetMapping("/invite/{inviteCode}")
    @Operation(summary = "초대 코드로 그룹 조회", description = "초대 코드로 그룹을 조회합니다")
    public ResponseEntity<ApiResponse<GroupDto>> getGroupByInviteCode(@PathVariable String inviteCode) {
        log.info("GET /api/v1/groups/invite/{} - Getting group by invite code", inviteCode);
        return groupService.getGroupByInviteCode(inviteCode)
                .map(group -> {
                    ApiResponse<GroupDto> response = ApiResponse.<GroupDto>builder()
                            .success(true)
                            .data(group)
                            .build();
                    return ResponseEntity.ok(response);
                })
                .orElseGet(() -> {
                    ApiResponse<GroupDto> response = ApiResponse.<GroupDto>builder()
                            .success(false)
                            .message("유효하지 않은 초대 코드입니다")
                            .build();
                    return ResponseEntity.status(HttpStatus.NOT_FOUND).body(response);
                });
    }

    @GetMapping("/{id}/members/count")
    @Operation(summary = "그룹 멤버 수 조회", description = "그룹의 활성 멤버 수를 조회합니다")
    public ResponseEntity<ApiResponse<Map<String, Long>>> countActiveMembers(@PathVariable String id) {
        log.info("GET /api/v1/groups/{}/members/count - Counting active members", id);
        long count = groupService.countActiveMembers(id);
        Map<String, Long> result = Map.of("count", count);
        ApiResponse<Map<String, Long>> response = ApiResponse.<Map<String, Long>>builder()
                .success(true)
                .data(result)
                .build();
        return ResponseEntity.ok(response);
    }

    @GetMapping("/check/name/{name}")
    @Operation(summary = "그룹 이름 중복 확인", description = "그룹 이름 중복 여부를 확인합니다")
    public ResponseEntity<ApiResponse<Map<String, Boolean>>> checkGroupName(@PathVariable String name) {
        log.info("GET /api/v1/groups/check/name/{} - Checking group name", name);
        boolean exists = groupService.existsByName(name);
        Map<String, Boolean> result = Map.of("exists", exists);
        ApiResponse<Map<String, Boolean>> response = ApiResponse.<Map<String, Boolean>>builder()
                .success(true)
                .data(result)
                .build();
        return ResponseEntity.ok(response);
    }
}
